syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.piggybank.sh.generated";

package sh.generated;

import "General.proto";

message Location
{
    int32 id = 1;

    GeoPoint geoPoint = 2;
}

message Event
{
    int32 id = 1;

    Location location = 2;
    TimeWindow timeWindow = 3;
    int32 duration = 4;
}

message Demand
{
    repeated Event events = 1;
}

message Order
{
    int32 id = 1;

    repeated Demand demands = 2;
}

message Task
{
    int32 id = 1;

    SearchParams params = 2;
    repeated Order orders = 3;
}

message Action
{
    int32 eventId = 1;
    int32 timeStart = 2;
}

message Stats
{
    int32 timeSpent = 1;
    int32 distanceTraveled = 2;
}

message Trip
{
    int32 orderId = 1;
    GeoPoint start = 2;
    GeoPoint finish = 3;
    repeated Action actions = 4;
    Stats stats = 5;
}

message Recommendations
{
    repeated Trip recommendations = 1;
}

service RecommendationsSearcher
{
    rpc findRecommendations(Task) returns (Recommendations) {};
}
